File: Jump_game2_BFS.py
# BFS
# Time complexity : O(n^k)
# Space complexity : O(n^k)
# Leetcode : Time limit exceeded

from collections import deque
class Solution:
    def jump(self, nums: List[int]) -> int:
        
        # if length of the nums list is less than 2, return 0
        n = len(nums)
        jump = 0
        if n < 2:
            return 0
        
        # add the first index to the queue
        q = deque([0])
        
        # traverse until the queue is not empty
        while q:
            # get the size of the queue as we are making node of the minimum jumps
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                # pop the element from the queue and iterate over the indexes that can be reached from the current position
                for i in range(1,nums[curr]+1):
                    idx = i + curr
                    
                    # if reached the end, then return the jumps + 1
                    if idx >= n - 1:
                        return jump + 1
                    # append the other indexes into the queue
                    q.append(idx)
            
            # increment jump after each level
            jump += 1

File: JumpGame2_BFS.java
//Time Coplexity =O(N^N), a bit optimised because we are removing already calculated ones
//Space complexity = O(N) + O(N);
//BFS
//Method: Maintain a count value and add root's children to the queue and a set; when children value is nums.length -1, return the count(i.e the min jumps needed to reach last index)
class Solution {
    public int jump(int[] nums) {
        //base
        if(nums.length < 2)return 0;
        HashSet<Integer> set = new HashSet<>();
        Queue<Integer> q = new LinkedList<>();
        q.add(0); set.add(0);
        int count =0;
        while(!q.isEmpty()){
            int size = q.size();
            //maintain tree level
            for(int i =0; i < size; i++){
                int currIdx = q.poll();

                //check for each index the highest it can jump to
                for(int j =0 ; j<= nums[currIdx]&& currIdx +j < nums.length; j++){
                    //if last element is reach return count +1
                    if(currIdx + j >= nums.length -1) return count +1 ;

                    //if idx not processed before then add it to the queue
                    if(!set.contains(currIdx+j)){
                        q.add(currIdx+j);
                        set.add(currIdx + j);
                    }
                }

                // System.out.println(q);

            }
            //increase the count after completing one level of tree
            count++;
            // System.out.println(count);
        }
        return count;
    }
}